### PR TITLE
feat: warn about unsaved changes in editor

### PR DESCRIPTION
**Pull request by Raubzeug**
[Stand](https://nda.ya.ru/t/UoVjvLhZ79eZCM)

### Test Status: <span style="color: orange;">⚠️ FLAKY</span>
📊 [Full Report](https://ydb-platform.github.io/ydb-embedded-ui/1620/)

| Total | Passed | Failed | Flaky | Skipped |
|:-----:|:------:|:------:|:-----:|:-------:|
| 134 | 133 | 0 | 1 | 0 |

### Bundle Size: 🔺
Current: 66.06 MB | Main: 66.03 MB
Diff: +0.04 MB (0.06%)

⚠️ Bundle size increased. Please review.

<details>
<summary>ℹ️ CI Information</summary>

- Test recordings for failed tests are available in the full report.
- Bundle size is measured for the entire 'dist' directory.
- 📊 indicates links to detailed reports.
- 🔺 indicates increase, 🔽 decrease, and ✅ no change in bundle size.
</details>

### CODE REVIEW

**Comment by astandrik** [comment]
I suppose this gives the same effect https://github.com/ydb-platform/ydb-embedded-ui/pull/1621

**Comment by Raubzeug** [comment]
> I suppose this gives the same effect #1621

@astandrik not exactly. It should warn if you try to insert query from History (or Saved queries, or Top queries).
Also I really don't like default browser confirmation >_<.
And last but not least, I don't think this should be managed via `useEffect`... this should be triggered by user's action, not when some props or state change.

**Comment by astandrik** [comment]
Updated my solution to more straigthforward approach =)

File: src/utils/hooks/withConfirmation/useChangeInputWithConfirmation.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+export async function getConfirmation(): Promise<boolean> {
+    return await NiceModal.show(CONFIRMATION_DIALOG, {
+        id: CONFIRMATION_DIALOG,
+        caption: i18n('context_unsaved-changes-warning'),
+        textButtonApply: i18n('action_apply'),
+    });
+}
+
+export function changeInputWithConfirmation<T>(callback: (args: T) => void) {
```
Comment (astandrik) [inline src/utils/hooks/withConfirmation/useChangeInputWithConfirmation.ts:19]: I suppose thats not only about input, but about any callback action

useActionWithConfirmation ?

File: src/containers/Tenant/Query/SavedQueries/SavedQueries.tsx [hunk 2]
```diff
@@ -154,7 +160,12 @@ export const SavedQueries = ({changeUserInput}: SavedQueriesProps) => {
                         settings={QUERY_TABLE_SETTINGS}
                         emptyDataMessage={i18n(filter ? 'history.empty-search' : 'saved.empty')}
                         rowClassName={() => b('row')}
-                        onRowClick={(row) => onQueryClick(row.body, row.name)}
+                        onRowClick={async (row) =>
```
Comment (astandrik) [inline src/containers/Tenant/Query/SavedQueries/SavedQueries.tsx:163]: according to gravity-ui/react-data-table onRowClick is sync function
Does passing async function have purpose here?

**Review by astandrik — APPROVED** [review]

File: src/utils/hooks/withConfirmation/i18n/en.json [hunk 3]
```diff
@@ -0,0 +1,4 @@
+{
+  "action_apply": "Proceed",
+  "context_unsaved-changes-warning": "You have unsaved changes in this query. Do you want to proceed?"
```
Comment (artemmufazalov) [inline src/utils/hooks/withConfirmation/i18n/en.json:3]: It's better to write in the Query editor or tab, "this query" isn't very clear thing when I am already in diagnostics or some other not query tab (e.g. Top Queries, History or Saved)

Screenshot 2024-11-14 at 16 12 39

**Review by artemmufazalov — CHANGES_REQUESTED** [review]

**Review by artemmufazalov — APPROVED** [review]

**Pull request merged**